### ISSUE TITLE
Recommended/better way to check for deprecated methods?

### ISSUE THREAD

**Issue by jasonwjones** [siom79/japicmp#188]
I have a need to check for methods that have been deprecated. I can't find anything "simple" to accomplish this, but I was able to get it done by "reaching in" to the CtMethod and doing the check based on an attribute named Deprecated:

	public static boolean isMethodNowDeprecated(JApiMethod method) {
		String attribute = "Deprecated";
		Optional<CtMethod> newMethod = method.getNewMethod();

		if (newMethod.isPresent()) {
			AttributeInfo newMethodAttribute = newMethod.get().getMethodInfo().getAttribute(attribute);
			Optional<CtMethod> oldMethod = method.getOldMethod();
			if (newMethodAttribute != null) {
				if (oldMethod.isPresent()) {
					// there's a new method, an old/existing method, and the new
					// method contains the Deprecated attribute but the old one
					// doesn't
					AttributeInfo oldMethodAttribute = oldMethod.get().getMethodInfo().getAttribute(attribute);
					return oldMethodAttribute == null;
				} else {
					// this can potentially happen if, for some reason,
					// the codebase adds a new method and deprecates it all in
					// the same release. 
					//
					// Or I suppose this could happen when you are comparing an
					// old JAR to a much newer one, where you are missing some
					// of the inbetween JARs, which had originally introduced
					// the new, non-deprecated method, and the newer JAR you are
					// looking at has already deprecated it
					return true;
				}
			}
		}
		// a method was removed, therefore it can't have been deprecated. Also
		// catches fall-through from above where the new method isn't deprecated
		// at all
		return false;
	}

**Comment by siom79** [comment]
Wouldn't you use the `AnnotationsAttribute` as described [here](https://jboss-javassist.github.io/javassist/html/javassist/bytecode/AnnotationsAttribute.html)?

**Comment by jasonwjones** [comment]
First of all, thanks for the link. I didn't realize that there were some static Strings that could be passed in to `getAttribute(String)` besides a specific annotation name. 

But my question is more along the lines of if it makes sense for the japicmp API to detect attribute changes in a way that is consistent with how it detects other changes. For example, it seems like it'd be good if I could ask a `JApiMethod` object about its attribute changes. That said, the `JApiMethod` object *does* have a `getAttributes()` method that returns `List<JApiAttribute<? extends Enum<?>>>` but this method *doesn't* seem to detect when the attribute is `@Deprecated` – and possibly some other attributes like `@Override`. My thinking at this time is that maybe this method detects arbitrary annotation changes (like if they were my own annotations) but not when they are essentially 'system' annotations. 

I have specifically compared two JARs with japicmp where the only difference was the addition of the `@Deprecated` attribute to a method, and the `getAttributes()` method on `JApiMethod` returns an empty list.

**Comment by siom79** [comment]
The point here is that annotating a method with @Deprecated does not change
it in the sense of an API change. But yes, it would be nice to list all
methods that have now been marked as deprecated.

Btw: Have you set the option `noAnnotations` to true? Otherwise japicmp
should also output changed annotations. Maybe you have a corner case in
which the annotations change is not reported correctly.

Jason Jones <notifications@github.com> schrieb am Di., 12. Dez. 2017, 23:14:

**Comment by jasonwjones** [comment]
Yes, I found that `noAnnotations` options and at first thought that might be the culprit, but it didn't make any difference (I tried with false and true). When I dug into the underlying `CtMethod` object I noticed that it seemed to have specific handling for Override/Deprecated, which led me to believe that maybe it was treating those as special annotations rather than 'user space' annotations. And you're right, these annotations don't really indicate an API change, but it would be nice to see them. The code I posted in the original comment gets the job done for now but I think it'd be nice to pick them up on the japicmp side of the fence. I can try my hand at code to implement and a pull request, but I wanted to just check and make sure I wasn't missing something obvious.

**Comment by siom79** [comment]
Pull request are always welcome. ;-)

Maybe it makes sense to make the feature even more generic such that the
users can configure which annotations should be reported. But for a first
version @Deprecated might be enough.

Jason Jones <notifications@github.com> schrieb am Di., 12. Dez. 2017, 23:43:
